src/endpoint/s3/ops/s3_post_object_select.js

```diff
@@ -0,0 +1,206 @@
+    //no PAYLOAD (DATA ends)
+    Buffer.from([0xfe, 0x2c, 0xee, 0x99]) //MESSAGE CRC (crc of above)
+]);
+
+//same for headers of records message
+const records_message_headers = Buffer.concat([
+    Buffer.from([11]), //byte length of below header name
+    Buffer.from(':event-type'),
+    Buffer.from([7, 0, 7]), //type (7, string) of below header value
+    Buffer.from('Records'),
+    Buffer.from([13]), //again for content-type header
+    Buffer.from(':content-type'),
+    Buffer.from([7, 0, 24]),
+    Buffer.from('application/octet-stream'),
+    Buffer.from([13]), //again for message-type header
+    Buffer.from(':message-type'),
+    Buffer.from([7, 0, 5]),
+    Buffer.from('event')
+]);
+
+class S3SelectStream extends Transform {
+
+    constructor(opts) {
+        super(opts);
+        this.s3select = new (nb_native().S3Select)(opts);
+    }
+
+    async encode_chunk(select_result) {
+        dbg.log2("select res = ", select_result.toString());
+
+        const total_length_bytes = select_result.length + records_message_headers.length + 16;
+        const buffer = Buffer.alloc(total_length_bytes);
+        let index = 0;
+
+        //prelude:
+        index = buffer.writeUInt32BE(total_length_bytes);
+        index = buffer.writeUInt32BE(records_message_headers.length, index);
+        index = buffer.writeInt32BE(crc32.buf(buffer.slice(0, index)), index); //prelude crc
+        //headers:
+        index += records_message_headers.copy(buffer, index);
+        //body:
+        index += select_result.copy(buffer, index);
+        //message crc:
+        buffer.writeInt32BE(crc32.buf(buffer.slice(0, index)), index);
+
+        //dbg.log2("select content response = ", buffer.toString('hex').match(/../g).join(' '));
+        return buffer;
+    }
+
+    async handle_result(result, cb) {
+        if (result.length > 0) {
+            this.push(await this.encode_chunk(result));
+        }
+
+        return cb();
+    }
+
+    async _transform(chunk, encoding, cb) {
+        try {
+            const select_result = await this.s3select.write(chunk);
+            return this.handle_result(select_result, cb);
+        } catch (err) {
+            dbg.error(err);
+            return cb(err);
+        }
+    }
+
+    async _flush(cb) {
+        try {
+            const select_result = await this.s3select.flush();
+            return this.handle_result(select_result, cb);
+        } catch (err) {
+            dbg.error(err);
+            return cb(err);
+        }
+    }
+}
+
+class EchoDebugStream extends Transform {
+
+    constructor(opts) {
+        super(opts);
+        this.header = opts.header;
+    }
+
+    _transform(chunk, enc, cb) {
+        dbg.log2(this.header, chunk.toString());
+        this.push(chunk);
+        return cb();
+    }
+}
+
+/**
+ * https://docs.aws.amazon.com/AmazonS3/latest/API/API_SelectObjectContent.html
+ */
+// eslint-disable-next-line max-statements
+async function post_object_select(req, res) {
```

Comment:
I think this function should apear at the begining of the file.

Reply:
@alphaprinz, Would you please move these classes to [s3_utils](https://github.com/noobaa/noobaa-core/blob/master/src/endpoint/s3/s3_utils.js) and just require at the top as @guymguym suggested? (I would like your main function to appear at the beginning of the file).